sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala

```diff
@@ -0,0 +1,140 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.hive.execution
+
+import java.io.IOException
+
+import scala.collection.JavaConverters._
+
+import org.apache.hadoop.hive.common.StatsSetupConst
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.analysis.CastSupport
+import org.apache.spark.sql.catalyst.catalog.{CatalogStatistics, CatalogTable, CatalogTablePartition, HiveTableRelation}
+import org.apache.spark.sql.catalyst.expressions.{And, AttributeReference, AttributeSet, BindReferences, Expression, Literal, SubqueryExpression}
+import org.apache.spark.sql.catalyst.planning.PhysicalOperation
+import org.apache.spark.sql.catalyst.plans.logical.{Filter, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.command.CommandUtils
+import org.apache.spark.sql.hive.client.HiveClientImpl
+import org.apache.spark.sql.types.BooleanType
+
+/**
+ * TODO: merge this with PruneFileSourcePartitions after we completely make hive as a data source.
+ */
+private[sql] class PruneHiveTablePartitions(session: SparkSession)
+  extends Rule[LogicalPlan] with CastSupport {
+
+  override val conf = session.sessionState.conf
+
+  /**
+   * Extract the partition filters from the filters on the table.
+   */
+  private def extractPartitionPruningFilters(
+              filters: Seq[Expression],
+              relation: HiveTableRelation): Seq[Expression] = {
+    val normalizedFilters = filters.map { e =>
+      e transform {
+        case a: AttributeReference =>
+          a.withName(relation.output.find(_.semanticEquals(a)).get.name)
+      }
+    }
+    val partitionSet = AttributeSet(relation.partitionCols)
+    normalizedFilters.filter { predicate =>
+      !predicate.references.isEmpty && predicate.references.subsetOf(partitionSet)
+    }
+  }
+
+  /**
+   * Prune the hive table using filters on the partitions of the table.
+   */
+  private def prunePartitions(relation: HiveTableRelation, partitionFilters: Seq[Expression])
+  : Seq[CatalogTablePartition] = {
+    val partitions =
+    if (conf.metastorePartitionPruning) {
+      session.sessionState.catalog.listPartitionsByFilter(
+        relation.tableMeta.identifier, partitionFilters)
+    } else {
+      session.sessionState.catalog.listPartitions(relation.tableMeta.identifier)
+    }
+    val shouldKeep = partitionFilters.reduceLeftOption(And).map { filter =>
```

Comment:
do we need this? AFAIK `ExternalCatalog.listPartitionsByFilter` does it already.

Reply:
Like in HiveTableScanExec, it calls prunePartitions(rawPartitions) in doExecute, while the rawPartitions already call ExternalCatalog.listPartitionsByFilter itself if HIVE_METASTORE_PARTITION_PRUNING enabled. 

It push the partition filters down into hive metastore for earlier pruning, but it still prune it again in prunePartitions, because some predicates may not be supported in hive metastore, like "b like 'xyz'".

So i think it is necessary here, WDYT?

Reply:
you are right hive may not support some predicates, so `HiveClientImpl.listPartitionByFilter` may return more partitions than we expect. But it should be fixed in `HiveExternalCatalog.listPartitionsByFilter`

Reply:
yea, make sense. What about creating a new PR for it and update it in both PruneHiveTablePartitions and HiveTableScanExec after it finished ?

Reply:
I mean, it's already fixed, `ExternalCatalog.listPartitionsByFilter` guarantees that the returned partitions are exactly what we want.

Reply:
yes, it is. Just checked.
Then the HiveTableScanExec should also be updated. right?

Reply:
Yea, but we can leave it in followup. The code newly added here should be updated.

Reply:
yea, updated.